identity/app/controllers/AccountDeletionController.scala

```diff
@@ -0,0 +1,162 @@
+          case FailedToEnterCorrectPassword =>
+            SeeOther(routes.AccountDeletionController.renderAccountDeletionForm ().url)
+              .flashing(boundForm.withError ("password", "Password is incorrect").toFlash)
+
+          case FailedToUnauthenticateUser =>
+            logger.error(s"Failed to un-authenticate user ${request.user.user.id} during account deletion.")
+            renderFormWithUnableToDeleteAccountError(boundForm)
+
+          case FailedToRemoveFromAllEmailLists =>
+            logger.error(s"Failed to remove user ${request.user.user.id} from all mailing lists during account deletion.")
+            renderFormWithUnableToDeleteAccountError(boundForm)
+
+          case FailedToDeleteAccount =>
+            logger.error(s"Failed to delete account for user ${request.user.user.id}")
+            renderFormWithUnableToDeleteAccountError(boundForm)
+        }
+
+      (for {
+        _ <- checkUserEnteredCorrectPassword()
+        _ <- removeFromAllMalingLists()
+        _ <- unauthenticateUser()
+        _ <- deleteAccountProper()
+      } yield ()).fold(processFailures, _ => clearCookiesAndDisplaySuccessForm)
+    }
+
+    sealed trait DeletionType
+    case object AutoDeletion extends DeletionType
+    case class ManualDeletion(hasSavedArticles: Boolean, hasComment: Boolean, hasJob: Boolean, isSubscriber: Boolean, isMember: Boolean) extends DeletionType
+
+
+    /* Users can delete account themselves if they
+         - has no saved articles, and
+         - have never commented, and
+         - do not have a jobs account, and
+         - do not have active digipack subscriptions, and
+         - do not have active membership */
+    private def selectDeletionType[A](implicit request: AuthenticatedActions.AuthRequest[A]): Future[DeletionType] =
+      for {
+        hasSavedArticles <- identityApiClient.hasNonDefaultSavedArticles(request.user.auth)
+        hasComment <- discussionApi.myProfile(request.headers).map { _.privateFields.fold(true)(_.hasCommented) }
+        hasJob <- Future.successful(request.user.user.userGroups.find(_.packageCode == "GRS").fold(false)(_ => true))
+        isSubscriber <- identityApiClient.userIsSubscriber(request.user.auth)
+        isMember <- identityApiClient.userIsMember(request.user.auth)
+      } yield {
+        if (!hasSavedArticles && !hasComment && !hasJob && !isSubscriber && !isMember)
+          AutoDeletion
+        else
+          ManualDeletion(hasSavedArticles, hasComment, hasJob, isSubscriber, isMember)
+      }
+
+    private def renderFormWithUnableToDeleteAccountError(boundForm: Form[String]) =
+      SeeOther(routes.AccountDeletionController.renderAccountDeletionForm ().url)
+        .flashing(boundForm.withGlobalError(
+          "We are experiencing technical difficulties. Your account has not been deleted. Please try again later or contact Userhelp.").toFlash)
+
+    private val cookiesToDiscard = List(
```

Comment:
There *must* already be a logout endpoint?! Why duplicate this hard-coded list of cookie names?